src/rust/engine/src/tasks.rs

```diff
@@ -45,43 +45,106 @@ impl rule_graph::Rule for Rule {
       &Rule::Intrinsic(_) => false,
     }
   }
+
+  fn color(&self) -> Option<rule_graph::Palette> {
+    match self {
+      Rule::Task(_) => None,
+      Rule::Intrinsic(_) => Some(rule_graph::Palette::Gray),
+    }
+  }
+
+  fn fmt_for_graph(&self) -> String {
+    match self {
+      Rule::Task(ref task) => {
+        let TaskElements {
+          rule_type,
+          task_name,
+          clause_portion,
+          product,
+          get_portion,
+        } = Self::extract_task_elements(task, 2, 1);
+
+        format!(
+          "@{}({}) -> {}{}\n{}",
+          rule_type, clause_portion, product, get_portion, task_name,
+        )
+      }
+      Rule::Intrinsic(ref intrinsic) => format!(
+        "@rule(<intrinsic>({}) -> {})",
+        intrinsic.input, intrinsic.product,
+      ),
+    }
+  }
+}
+
+struct TaskElements {
+  rule_type: String,
+  task_name: String,
+  clause_portion: String,
+  product: String,
+  get_portion: String,
+}
+
+impl Rule {
+  fn extract_task_elements(
+    task: &Task,
+    select_clause_threshold: usize,
+    get_clause_threshold: usize,
+  ) -> TaskElements {
+    let product = format!("{}", task.product);
+    let select_clauses = task
+      .clause
+      .iter()
+      .map(|c| c.product.to_string())
+      .collect::<Vec<_>>();
+    let clause_portion = if select_clauses.len() <= select_clause_threshold {
+      select_clauses.join(", ").to_string()
+    } else {
+      format!("\n{},\n", select_clauses.join(",\n"))
+    };
+    let get_clauses = task
+      .gets
+      .iter()
+      .map(::std::string::ToString::to_string)
+      .collect::<Vec<_>>();
+    let rule_type = if task.cacheable {
+      "rule".to_string()
+    } else {
+      "goal_rule".to_string()
+    };
+    let get_portion = if get_clauses.is_empty() {
+      "".to_string()
+    } else if get_clauses.len() <= get_clause_threshold {
+      format!(",\ngets=[{}]", get_clauses.join(", "))
```

Comment:
Can you please remove the `\n` portion in this case? It makes a lot of sense for the visualization, but I don't think we want it for rule graph errors due to the reasons explained in prior comments. 

It looks like this is the sole thing we'd need to change to avoid impacting rule graph errors at all, which attests to how neat this abstraction is!